src/measure_theory/prod.lean

```diff
@@ -0,0 +1,689 @@
+begin
+  have m := @measurable_prod_mk_left,
+  refine measurable.ennreal_induction _ _ _,
+  { intros c s hs, simp only [← indicator_comp_right],
+    simp [lintegral_indicator, m hs, hs, lintegral_const_mul, measurable_measure_prod_mk_left hs,
+      prod_apply] },
+  { rintro f g - hf hg h2f h2g,
+    simp [lintegral_add, measurable.lintegral_prod_right', hf.comp m, hg.comp m,
+      hf, hg, h2f, h2g] },
+  { intros f hf h2f h3f,
+    have kf : ∀ x n, measurable (λ y, f n (x, y)) := λ x n, (hf n).comp m,
+    have k2f : ∀ x, monotone (λ n y, f n (x, y)) := λ x i j hij y, h2f hij (x, y),
+    have lf : ∀ n, measurable (λ x, ∫⁻ y, f n (x, y) ∂ν) := λ n, (hf n).lintegral_prod_right',
+    have l2f : monotone (λ n x, ∫⁻ y, f n (x, y) ∂ν) := λ i j hij x, lintegral_mono (k2f x hij),
+    simp only [lintegral_supr hf h2f, lintegral_supr (kf _), k2f, lintegral_supr lf l2f, h3f] },
+end
+
+/-- The symmetric verion of Tonelli's Theorem: For `ennreal`-valued measurable functions on `α × β`,
+  the integral of `f` is equal to the iterated integral, in reverse order. -/
+lemma lintegral_prod_symm [sigma_finite μ] (f : α × β → ennreal)
+  (hf : measurable f) : ∫⁻ z, f z ∂(μ.prod ν) = ∫⁻ y, ∫⁻ x, f (x, y) ∂μ ∂ν :=
+by { simp_rw [← lintegral_prod_swap f hf], exact lintegral_prod _ (hf.comp measurable_swap) }
+
+/-- The reversed version of Tonelli's Theorem. In this version `f` is in curried form, which makes
+  it easier for the elaborator to figure out `f` automatically. -/
+lemma lintegral_lintegral ⦃f : α → β → ennreal⦄
+  (hf : measurable (uncurry f)) :
+  ∫⁻ x, ∫⁻ y, f x y ∂ν ∂μ = ∫⁻ z, f z.1 z.2 ∂(μ.prod ν) :=
+(lintegral_prod _ hf).symm
+
+/-- The reversed version of Tonelli's Theorem (symmetric version). In this version `f` is in curried
+  form, which makes it easier for the elaborator to figure out `f` automatically. -/
+lemma lintegral_lintegral_symm [sigma_finite μ] ⦃f : α → β → ennreal⦄
+  (hf : measurable (uncurry f)) :
+  ∫⁻ x, ∫⁻ y, f x y ∂ν ∂μ = ∫⁻ z, f z.2 z.1 ∂(ν.prod μ) :=
+(lintegral_prod_symm _ (hf.comp measurable_swap)).symm
+
+/-- Change the order of Lebesgue integration. -/
+lemma lintegral_lintegral_swap [sigma_finite μ] ⦃f : α → β → ennreal⦄
+  (hf : measurable (uncurry f)) :
+  ∫⁻ x, ∫⁻ y, f x y ∂ν ∂μ = ∫⁻ y, ∫⁻ x, f x y ∂μ ∂ν :=
+(lintegral_lintegral hf).trans (lintegral_prod_symm _ hf)
+
+/-! ### Integrability on a product -/
+section
+
+variables [opens_measurable_space E]
+
+lemma integrable.swap [sigma_finite μ] ⦃f : α × β → E⦄
+  (hf : integrable f (μ.prod ν)) : integrable (f ∘ prod.swap) (ν.prod μ) :=
+⟨hf.measurable.comp measurable_swap,
+  (lintegral_prod_swap _ hf.measurable.ennnorm : _).le.trans_lt hf.has_finite_integral⟩
+
+lemma integrable_swap_iff [sigma_finite μ] ⦃f : α × β → E⦄ :
+  integrable (f ∘ prod.swap) (ν.prod μ) ↔ integrable f (μ.prod ν) :=
+⟨λ hf, by { convert hf.swap, ext ⟨x, y⟩, refl }, λ hf, hf.swap⟩
+
+lemma has_finite_integral_prod_iff ⦃f : α × β → E⦄ (h1f : measurable f) :
```

Comment:
The name suggest that the statement is `has_finite_integral f (μ.prod ν) ↔ ...`

Reply:
I could make the statement that (violating "RHS is simpler"). I'm not sure what a good name is otherwise. `iff_has_finite_integral_prod` has a non-useful prefix, and describing LHS takes too long. Suggestions?

Reply:
This is not going to be a `simp` lemma so I think that "RHS is simpler" is not a strict rule here.